src/main/java/cn/nukkit/utils/BinaryStream.java

```diff
@@ -491,33 +497,37 @@ public void putSlot(Item item) {
     }
 
     public void putSlot(Item item, boolean instanceItem) {
-        if (item == null || item.getId() == 0) {
+        if (item == null || item.getId() == Item.AIR) {
             this.putByte((byte) 0);
             return;
         }
 
+        int id = item.getId();
+        int meta = item.getDamage();
+        boolean isBlock = item instanceof ItemBlock;
+        boolean isDurable = item instanceof ItemDurable;
+
         RuntimeItemMapping mapping = RuntimeItems.getMapping();
-        RuntimeEntry runtimeEntry = mapping.toRuntime(item.getId(), item.getDamage());
+        RuntimeEntry runtimeEntry = mapping.toRuntime(id, meta);
         int runtimeId = runtimeEntry.getRuntimeId();
-        int damage = runtimeEntry.isHasDamage() ? 0 : item.getDamage();
+        int damage = isBlock || isDurable || runtimeEntry.isHasDamage() ? 0 : meta;
 
         this.putVarInt(runtimeId);
         this.putLShort(item.getCount());
-
         this.putUnsignedVarInt(damage);
 
         if (!instanceItem) {
             this.putBoolean(true);
             this.putVarInt(1);
         }
 
-        Block block = item.getBlockUnsafe();
+        Block block = isBlock ? item.getBlockUnsafe() : null;
         int blockRuntimeId = block == null ? 0 : GlobalBlockPalette.getOrCreateRuntimeId(block.getId(), block.getDamage());
         this.putVarInt(blockRuntimeId);
 
         ByteBuf userDataBuf = ByteBufAllocator.DEFAULT.ioBuffer();
         try (LittleEndianByteBufOutputStream stream = new LittleEndianByteBufOutputStream(userDataBuf)) {
-            if (((item instanceof ItemDurable && item.getDamage() > 0) || block != null && block.getDamage() > 0) && !runtimeEntry.isHasDamage()) {
```

Comment:
Similar here. If item holds its damage/meta in runtime id, do not sent it in NBT.